benchmark/app.py

```diff
@@ -0,0 +1,94 @@
+import os
+import time
+import uuid
+from math import ceil
+
+import click
+from pandas import DataFrame
+
+from jina.flow import Flow
+from jina.drivers import BaseDriver
+
+from post_process import clean_dataframe, evaluate_times, get_summary
+from utils import bytes_generator, generate_filename, get_list_of_num_docs
+from summary import html_table, plot_num_docs_vs_time
+
+FILE_DIR = os.environ.get('FILE_DIR', './.data')
+os.makedirs(FILE_DIR, exist_ok=True)
+
+START_NUM_DOCS = 512
+MULTIPLIER_NUM_DOCS = 4
+COUNT_NUM_DOCS = 10
+DEFAULT_BATCH_SIZE = 256
+COLUMN_OF_INTEREST = 'g:send' # to be changed to 'roundtrip' once issue is fixed
+
+
+class BenchmarkDriver(BaseDriver):
+    routes = []
+
+    def __init__(self, num_pods: str = '2', num_epochs: str = '2', filename: str = 'routes.parquet', *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.num_pods = int(num_pods)
+        self.num_epochs = int(num_epochs)
+        self.filename = filename
+
+    def __call__(self, *args, **kwargs):
+        if (len(self.envelope.routes) == self.num_pods + 1) and (len(self.req.docs) > 0):
+            # 1st condition is to make sure we extract `routes` only at the last pod
+            # 2nd condition is to make sure we don't write to the file during dry_run
+            current_route_dict = {}
+            for _pod in self.envelope.routes:
+                name = _pod.pod
+                start_time = _pod.start_time.ToDatetime()
+                end_time = _pod.end_time.ToDatetime()
+                current_route_dict[name] = [start_time, end_time]
+            BenchmarkDriver.routes.append(current_route_dict)
+            if len(BenchmarkDriver.routes) == self.num_epochs:
+                DataFrame(BenchmarkDriver.routes).to_parquet(f'{FILE_DIR}/{self.filename}')
+
+
+@click.command()
+@click.option('--num-bytes', '-b', default=10)
+@click.option('--num-pods', default=2)
+@click.option('--batch-size', default=DEFAULT_BATCH_SIZE)
+def run_benchmark(num_bytes, num_pods, batch_size):
```

Comment:
Great work! 

One thing I am missing the point of just benchmarking a Flow with 2 Pods that do nothing. I feel like it could be more interesting to benchmark some indexing flow and see later its query flow execution time?. I think the issue that lead to this story had this motivation in mind actually.

Reply:
Maybe we could let the user provide any Flow yaml comnfiguration and force in this app to include a BenchMarkDriver (in the core I think there are some Profiling Drivers although not sure if used). Or even assume that the user will have added this Driver (at the end it would be mostly used internally)?

Reply:
It makes sense to add some functionality in the pods to check the user's issue. 
But for the raw benchmarking, which we'll compare across different Jina releases, can remain vague, don't you think The inspiration is from https://github.com/hanxiao/benchmark/tree/master. We can link that to a Github action for an automatic trigger.

Reply:
I would make the test Flow agnostic and have by default this Flow yaml with the driver. But then it can be used in many ways to check really often used flow archs.

For thr github action, i think is not a good idea to have perf measures from github actions since we have no control on hardware or occupancy of the github machines. I think this can lead to false information. I believe we had this conversation some time ago with @Nan Wang.